### DISCUSSION TITLE
Error: Can't find model 'en_core_web_sm'. It doesn't seem to be a Python package or a valid path to a data directory

### DISCUSSION THREAD

**Discussion by simonrmortensen** [explosion/spaCy#10895] [Help: Installation]
I'm trying to load the ```en_core_web_sm``` spaCy model, but I have been unsuccessful in doing so. 

The error that occurs is the following: 

```
OSError: [E050] Can't find model 'en_core_web_sm'. It doesn't seem to be a Python package or a valid path to a data directory.
```

I'm working in a Anaconda virtual environment. The following checkboxes are ticked: 

 - Did ```conda activate gcp-env``` prior to installing spaCy and the english language model
 - Have run ```conda install -c conda-forge spacy``` while on the right environment
 - Then, have run ```python -m spacy download en```, still while on the right environment
 - Also tried adding ```spacy``` to the ```requirements.txt``` , and installing dependencies via that route, after first attempts failed

```spacy info``` produces this output: 
```
spacy info

============================== Info about spaCy ==============================

spaCy version    3.3.0                         
Location         /Users/simonmortensen/opt/anaconda3/envs/gcp-env/lib/python3.10/site-packages/spacy
Platform         macOS-11.6.5-x86_64-i386-64bit
Python version   3.10.4                        
Pipelines        en_core_web_sm (3.3.0)
```

```python -m spacy validate``` produces this output:
````
================= Installed pipeline packages (spaCy v3.3.0) =================
ℹ spaCy installation:
/Users/simonmortensen/opt/anaconda3/envs/gcp-env/lib/python3.10/site-packages/spacy

NAME             SPACY                 VERSION                            
en_core_web_sm   >=3.3.0.dev0,<3.4.0   3.3.0   ✔
````

For additional context, ```pip list``` on the environment contains both 
```
spacy                         3.3.0
spacy-legacy                  3.0.9
spacy-loggers                 1.0.2
```
and 
```
en-core-web-sm                3.3.0
```
Even so, ```import en_core_web_sm``` also doesn't work: 
```
import en_core_web_sm
Traceback (most recent call last):

  Input In [65] in <cell line: 1>
    import en_core_web_sm

ModuleNotFoundError: No module named 'en_core_web_sm'
```

I've been through several previous StackOverflow posts on the same topic. Those have often been solved, but my issue remains. So I've also posted this same question to [StackOverflow](https://stackoverflow.com/questions/72463970/spacy-cant-find-model-en-core-web-sm-it-doesnt-seem-to-be-a-python-package), but have not been able to find an answer (so far...). 

Any advice would be very much appreciated. Thanks in advance!

Simon

**Comment by koaning** [comment]
This might be a virtual environment thing.  

There could be a difference between: 

```
python -m spacy download en_core_web_sm
```

and 

```
spacy download en_core_web_sm
```

I couldn't help but notice some of your commands omit `python -m`, which might cause some of the issues we're seeing. If you're curious about details, I've attempted to make a tutorial on what might go wrong [here](https://calmcode.io/virtualenv/bad-things.html). 

Could you try running: 

```bash
# Use python explicitly to grab info. 
python -m spacy info

# Just confirm the versions via pip too
python -m pip freeze | grep spacy

# Download the en_core_web_sm model explicitly 
python -m spacy download en_core_web_sm
``` 

Once that's downloaded, you should be able to load in the model via: 

```python
import spacy 

nlp = spacy.load("en_core_web_sm")
```

**Reply by simonrmortensen** [reply]
Hey @koaning, thanks for those video resources! I wasn't able to solve the issue with them in this case, but they certainly cleared a lot of things up for me regardless. 

Anyway, I ran the commands you suggested, and at the end of them, I still got the same error. 
```
(gcp-env) Simons-MBP:~ simonmortensen$ python -m spacy info

============================== Info about spaCy ==============================

spaCy version    3.3.0                         
Location         /Users/simonmortensen/opt/anaconda3/envs/gcp-env/lib/python3.10/site-packages/spacy
Platform         macOS-11.6.5-x86_64-i386-64bit
Python version   3.10.4                        
Pipelines        en_core_web_sm (3.3.0) 
```
```
(gcp-env) Simons-MBP:~ simonmortensen$ python -m pip freeze | grep spacy
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.3.0/en_core_web_sm-3.3.0-py3-none-any.whl
spacy @ file:///Users/runner/miniforge3/conda-bld/spacy_1651591371931/work
spacy-legacy @ file:///home/conda/feedstock_root/build_artifacts/spacy-legacy_1645713043381/work
spacy-loggers @ file:///home/conda/feedstock_root/build_artifacts/spacy-loggers_1648794027183/work
```

On ```(gcp-env) Simons-MBP:~ simonmortensen$ python -m spacy download en_core_web_sm``` the output started with ```Collecting en-core-web-sm==3.3.0```, and ended with ```✔ Download and installation successful. You can now load the package via spacy.load('en_core_web_sm')```, without errors, just as before. 

I did notice something when I typed ```python -m conda```. The output was ```/Users/simonmortensen/opt/anaconda3/envs/gcp-env/bin/python: No module named conda```. I wonder if the original ```conda install -c conda-forge spacy``` might be the root of the problem, because that's run on the global ```conda```. What do you think?

**Reply by koaning** [reply]
Is there a reason why you must run conda? I need to be honest; I've never really used conda before and so far my `pip` installations have always worked out fine. I don't own a Windows machine though, and my impression is that conda does make a lot of things much easier for that operating system. 

Maybe a completely new env via pip might be worth a try? 

```bash
# Start a new virtualenv
python3 -m venv venv 

# Activate the virtualenv 
source venv/bin/activate 

# Upgrade pip before installing anything, ever 
python -m pip install --upgrade pip 

# Install spaCy 
python -m pip install spacy 

# Download model 
python -m spacy download en_core_web_sm

# Run Python
python
```

From here, you should be able to run: 

```python
import spacy 

nlp = spacy.load("en_core_web_sm")
```

**Reply by simonrmortensen** [reply]
I use conda mainly because I'm more familiar with that, and have had some issues in the past by mixing conda with "non-conda". 

Regardless, I tried your suggestion above, creating and sourcing a virtual environment outside of anaconda. I did exactly the same steps as you suggested above (except I ran ```spyder``` in the end instead of ```python```, as that's my preferred IDE. Spyder is installed via ```conda``` though, would that be a problem?). 

I got the following output from the console in Spyder: 
```
!python -m pip freeze | grep spacy
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.3.0/en_core_web_sm-3.3.0-py3-none-any.whl
spacy==3.3.0
spacy-legacy==3.0.9
spacy-loggers==1.0.2

import spacy

nlp = spacy.load("en_core_web_en")
Traceback (most recent call last):

  Input In [15] in <cell line: 1>
    nlp = spacy.load("en_core_web_en")

  File ~/opt/anaconda3/lib/python3.9/site-packages/spacy/__init__.py:51 in load
    return util.load_model(

  File ~/opt/anaconda3/lib/python3.9/site-packages/spacy/util.py:427 in load_model
    raise IOError(Errors.E050.format(name=name))

OSError: [E050] Can't find model 'en_core_web_en'. It doesn't seem to be a Python package or a valid path to a data directory.
```

At this point I also got another (similar but maybe unrelated) issue: 
```
!python -m pip freeze | grep google-cloud
google-cloud==0.34.0
google-cloud-vision==2.7.2

from google.cloud import vision
Traceback (most recent call last):

  Input In [10] in <cell line: 1>
    from google.cloud import vision

ImportError: cannot import name 'vision' from 'google.cloud' (unknown location)
```

The error message is not the same, but the behaviour is rather similar, as I can't import/load modules even though I have proof of them being installed. Hopefully that provides some extra context?

This makes 0 sense to me. Have you got any ideas?

**Reply by koaning** [reply]
This is indeed a bit strange. But look at your log: 

```
  Input In [15] in <cell line: 1>
    nlp = spacy.load("en_core_web_en")

  File ~/opt/anaconda3/lib/python3.9/site-packages/spacy/__init__.py:51 in load
    return util.load_model(

  File ~/opt/anaconda3/lib/python3.9/site-packages/spacy/util.py:427 in load_model
    raise IOError(Errors.E050.format(name=name))
```

It seems that your Python env is located at `~/opt/anaconda3/lib/python3.9`? That feels like you're still using a virtualenv from `conda`. 

Did you activate your virtualenv? You can typically check via `which(python)` to understand where your virtualenv is located. 

Also ... I couldn't help but notice you're running your Python code from inside of a Jupyter notebook? Hence the `!` in `!python -m pip freeze | grep spacy`. Are you 100% sure that this Jupyter notebook is launched from the correct virtualenv? Did you start the notebook via `python -m jupyter lab`? 

The video series goes into a [bit of depth](https://calmcode.io/virtualenv/good-fixes.html) on what could all be going wrong here.

**Reply by simonrmortensen** [reply]
Good catch!

So I uninstalled ```spacy``` from anaconda, deactivated all ```conda``` environments, and ran the following commands to verify: 
````
simonrmortensen@MacBook-Pro-som-tillhor-Simon ~ % source gcp-env/bin/activate
(gcp-env) simonrmortensen@MacBook-Pro-som-tillhor-Simon ~ % which python
/Users/simonrmortensen/gcp-env/bin/python
(gcp-env) simonrmortensen@MacBook-Pro-som-tillhor-Simon ~ % which pip
/Users/simonrmortensen/gcp-env/bin/pip
(gcp-env) simonrmortensen@MacBook-Pro-som-tillhor-Simon ~ % which spacy
/Users/simonrmortensen/gcp-env/bin/spacy
````

However, now, I can't even ```import spacy```, let alone the english model: 
```
import spacy
Traceback (most recent call last):

  Input In [4] in <cell line: 1>
    import spacy

ModuleNotFoundError: No module named 'spacy'


!python -m pip freeze | grep spacy
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.3.0/en_core_web_sm-3.3.0-py3-none-any.whl
spacy==3.3.0
spacy-legacy==3.0.9
spacy-loggers==1.0.2
```

🤷🏼‍♂️

**Reply by koaning** [reply]
I once again notice the `!python -m pip freeze | grep spacy`. Are you running it from Jupyter or ipython? Are you 100% sure that this Jupyter/ipython is in the same virtualenv and not from conda?

**Reply by simonrmortensen** [reply]
I deactivated and removed the ```conda``` venv completely, and then ```source gcp-env/bin/activate```, before installing all the packages. 

And all the ```which ...``` commands confirm that the package was in the new python venv - no longer an /anaconda3/ folder. 

I'm clearly not too familiar with the terminal, nor python environments, so please correct me if that is a false assumption.

**Reply by koaning** [reply]
But you're running the Python command in Jupyter, right? The errors you mention talk about cells? 

```
Traceback (most recent call last):

  Input In [4] in <cell line: 1>
    import spacy
```

**Reply by simonrmortensen** [reply]
Oh, you meant the ```!``` in front of the command? Yes, that was run from the Spyder IDE. 

However, the terminal produces the same output:
```
(gcp-env) simonrmortensen@MacBook-Pro-som-tillhor-Simon ~ % python -m pip freeze | grep spacy
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.3.0/en_core_web_sm-3.3.0-py3-none-any.whl
spacy==3.3.0
spacy-legacy==3.0.9
spacy-loggers==1.0.2
(gcp-env) simonrmortensen@MacBook-Pro-som-tillhor-Simon ~ % python -m pip freeze | grep google-cloud
google-cloud==0.34.0
google-cloud-vision==2.7.2
```

**Reply by koaning** [reply]
Could you try running it from your terminal? At the moment I can't rule out that Spyder is doing something naughty.

**Reply by simonrmortensen** [reply]
You mean running the entire Python script in the terminal? 

Interesting output: 
```
(gcp-env) simonrmortensen@MacBook-Pro-som-tillhor-Simon ~ % python /Users/simonrmortensen/Desktop/testing.py
```
No errors (on ```import``` statements, at least). 

So yes, it does seem like Spyder is messing with it. Do you have any suggestions on how to troubleshoot?

**Reply by simonrmortensen** [reply]
[This site](https://docs.spyder-ide.org/current/faq.html#using-existing-environment) suggests that I change the Python interpreter in the Spyder preferences to the ```/bin/python``` in the new venv:
![Skärmavbild 2022-06-03 kl  13 50 09](https://user-images.githubusercontent.com/47061634/171848402-1a156163-f790-4dee-9a2a-1dee258ad2d1.png)

But this python file is just a pointer to the anaconda3 python file. So when I select it and restart the kernel, the problem still exists.

**Reply by koaning** [reply]
Could you try running the commands from a terminal? I understand it might be new to you, but today is as good a time as any to maybe get started with it.

**Reply by koaning** [reply]
Actually, because you said you were unfamiliar, you may appreciate following along with the entire course on [calmcode](https://calmcode.io/virtualenv/intro.html). I'm the maintainer of that project, and I don't want to toot my own horn here, but I can imagine just taking the time through all the steps there might also just help shape your understanding.

**Reply by simonrmortensen** [reply]
Happy to learn, and thanks for all the help! 😄 

I've run all commands inside the terminal, including ```python testing.py```. They all work (or at least, the ```import``` statements are no longer the error), so as you suggest, I'm assuming that Spyder is doing something funky. 

I'm researching right now, and at first glance it seems as though you actually need to install Spyder on every virtual environment, just like any other package. Trying this now.

**Reply by simonrmortensen** [reply]
As you suggested - Spyder was the villain. 

All packages were correctly installed on the virtual environment, but Spyder was not running that environment (even if the IDE was launched with the ```spyder``` command from a terminal where the environment was in fact activated). 

In order to make Spyder run the correct environment, you needed to change the Python interpreter in the Spyder preferences: 
<img width="1072" alt="Skärmavbild 2022-06-04 kl  09 59 42" src="https://user-images.githubusercontent.com/47061634/171990503-b4d8c0ba-ae12-45bb-8fbc-4eca5a751f56.png">
... and then restart the kernel. 

I got an error prompting me to ```pip install spyder-kernels==2.1.*```, but once that was done (make sure to do it on the right env) and I restarted Spyder, it finally worked! 

@koaning, thanks a ton for hopping on this thread, you've been a great help!

**Reply by koaning** [reply]
Happy to hear there's more calm now. 🙂

**Comment by ClaudioAlejandroTorre** [comment]
Hello everyone. I worked with nlp = spacy.load("en_core_web_sm") having generated .exe with PyInstaller, considering that in the python interpreter environment the software worked well. When executing the .exe the following error occurred: Error: Can't find model 'en_core_web_sm'. It doesn't seem to be a Python package or a valid path to a data directory. I was able to solve it by watching the video published in:

https://ar.video.search.yahoo.com/search/video;_ylt=AwrJ_GuHE1lmbjIAzFmr9Qt.;_ylu=Y29sbwNiZjEEcG9zAzEEdnRpZAMEc2VjA3BpdnM-?p=OSError%3A+%5BE050%5D+Can%27t+find+model+%27en_core_web_s m%27. +It+doesn%27t+seem+to+be+a+Python+package+or+a+valid+path+to+a+data+directory.&fr2=piv-web&type=E210AR739G0&fr=mcafee#id=2&vid=33eb7c89f0a4a4d380f2fb6865a978cc&action =view

It was necessary to replace the statement nlp = spacy.load("en_core_web_sm") in my source program with nlp = spacy.load("C:\Users\claud\OneDrive\Desktop\Python312\.venv\Lib\site-packages\en_core_web_sm \en_core_web_sm-3.7.1")
 nlp = spacy.load("C:/Users/claud/OneDrive/Desktop/Python312/.venv/Lib/site-packages/en_core_web_sm/en_core_web_sm-3.7.1"). 

That is, as the video explains, I directly replaced "en_core_web_sm" with the address on my Windows 11 notebook where the spacy library (python -m spacy download en). Thank you so much

**Reply by ClaudioAlejandroTorre** [reply]
Hello everyone. I worked with nlp = spacy.load("en_core_web_sm") having generated .exe with PyInstaller, considering that in the python interpreter environment the software worked well. When executing the .exe the following error occurred: Error: Can't find model 'en_core_web_sm'. It doesn't seem to be a Python package or a valid path to a data directory. I was able to solve it by watching the video published in: https://ar.video.search.yahoo.com/search/video;_ylt=AwrJ_GuHE1lmbjIAzFmr9Qt.;_ylu=Y29sbwNiZjEEcG9zAzEEdnRpZAMEc2VjA3BpdnM-?p=OSError%3A+%5BE050%5D+Can%27t+find+model+%27en_core_web_s m%27. +It+doesn%27t+seem+to+be+a+Python+package+or+a+valid+path+to+a+data+directory.&fr2=piv-web&type=E210AR739G0&fr=mcafee#id=2&vid=33eb7c89f0a4a4d380f2fb6865a978cc&action =view

It was necessary to replace the statement nlp = spacy.load("_en_core_web_sm_") in my source program with nlp = spacy.load("C:\Users\claud\OneDrive\Desktop\Python312\.venv\Lib\site-packages\en_core_web_sm \en_core_web_sm-3.7.1")
 nlp = spacy.load("C:/Users/claud/OneDrive/Desktop/Python312/.venv/Lib/site-packages/en_core_web_sm/en_core_web_sm-3.7.1"). That is, as the video explains, I directly replaced "en_core_web_sm" with the address on my Windows 11 notebook where the spacy library (python -m spacy download en). The .exe run ok. Thank you so much

**Reply by XYLxiria** [reply]
Now my code run perfectly! Thanks, i took solution#3 and now there is no bug about spacy

**Reply by kd65541** [reply]
referring to solution 3, you can also copy paste the en_core_web_sm folders from the installed path (stated in cmd prompt) to the path your IDE said was missing
